frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -770,46 +765,39 @@ function ReaderHighlight:onTapXPointerSavedHighlight(ges)
     --       i.e., often a bit earlier)...
+    -- Even in page mode, it's safer to use pos and ui.dimen.h
+    -- than pages' xpointers pos, even if ui.dimen.h is a bit
+    -- larger than pages' heights
+    local cur_view_top = self.document:getCurrentPos()
+    local cur_view_bottom
+    if self.view.view_mode == "page" and self.document:getVisiblePageCount() > 1 then
+        cur_view_bottom = cur_view_top + 2 * self.ui.dimen.h
+    else
+        cur_view_bottom = cur_view_top + self.ui.dimen.h
+    end
     local highlights_tapped = {}
-    for page, items in pairs(self.view.highlight.saved) do
-        if items then
-            for i = 1, #items do
-                local item = items[i]
-                local pos0, pos1 = item.pos0, item.pos1
-                -- document:getScreenBoxesFromPositions() is expensive, so we
-                -- first check this item is on current page
-                if not cur_view_top then
-                    -- Even in page mode, it's safer to use pos and ui.dimen.h
-                    -- than pages' xpointers pos, even if ui.dimen.h is a bit
-                    -- larger than pages' heights
-                    cur_view_top = self.ui.document:getCurrentPos()
-                    if self.view.view_mode == "page" and self.ui.document:getVisiblePageCount() > 1 then
-                        cur_view_bottom = cur_view_top + 2 * self.ui.dimen.h
-                    else
-                        cur_view_bottom = cur_view_top + self.ui.dimen.h
-                    end
-                end
-                local spos0 = self.ui.document:getPosFromXPointer(pos0)
-                local spos1 = self.ui.document:getPosFromXPointer(pos1)
-                local start_pos = math.min(spos0, spos1)
-                local end_pos = math.max(spos0, spos1)
-                if start_pos <= cur_view_bottom and end_pos >= cur_view_top then
-                    local boxes = self.ui.document:getScreenBoxesFromPositions(pos0, pos1, true) -- get_segments=true
-                    if boxes then
-                        for index, box in pairs(boxes) do
-                            if inside_box(pos, box) then
-                                logger.dbg("Tap on highlight")
-                                if self.select_mode then
-                                    if page == self.highlight_page and i == self.highlight_idx then
-                                        -- tap on the first fragment: abort select mode, clear highlight
-                                        self.select_mode = false
-                                        self:deleteHighlight(page, i)
-                                        return true
-                                    end
-                                else
-                                    table.insert(highlights_tapped, {page, i})
-                                    break
+    for hl_i, item in ipairs(self.ui.annotation.annotations) do
+        if item.drawer then
```

Comment:
And if this kind of iterating could be shove into a `ReaderAnnotations:iterateAnnotations(filter_func)` (there is some (ugly) Lua coroutine.yield() stuff to allow keeping having a single loop, ie. in https://github.com/koreader/koreader/blob/ad6607cb4c827f5e0a14a5933e6ac5dcbf22a271/frontend/util.lua#L78)

Reply:
I'm thinking about caching screen boxes, will consider ^^
https://github.com/koreader/koreader/blob/ad6607cb4c827f5e0a14a5933e6ac5dcbf22a271/frontend/apps/reader/modules/readerview.lua#L582-L585